README.md
# Analytics For Cloudflare

![logo](https://raw.githubusercontent.com/ChuckMac/analytics-for-cloudflare/assets/banner-772x250.jpg)

This is a WordPress plugin to connect your WordPress dashboard to your CloudFlare account to display some key analytics data.

The dashboard widget provides a quick look at some key metrics for your site.

 * Requests
 * Page Views
 * Unique Visitors
 * Bandwidth

The data can be viewed from several timeframes.

  * Last week
  * Last Month
  * Last 24 Hours

You can also get a quick look at some other metics available.

  * SSL vs Non-SSL
  * Breakdown of content types (html, jpg, png, js, etc..)
  * Breakdown of requests by country

## Minimum Requirements

* WordPress 3.8 or greater
* CloudFlare account (which is utilizing the CloudFlare network and not passthru)

## Installation Instructions

1. Upload the entire 'analytics-for-cloudflare' folder to the '/wp-content/plugins/' directory
2. Activate the plugin through the 'Plugins' menu in WordPress
3. Enter your CloudFlare API credentials in the 'Settings' -> 'Analytics For Cloudflare' screen
4. Once connected, select the domain to display the statics for on the same settings page.

## Screenshots

![Screenshot 1](https://raw.githubusercontent.com/ChuckMac/analytics-for-cloudflare/assets/screenshot-1.png)


CHANGELOG.md
# Analytics For Cloudflare

## 1.1 - 2017-02-09
### Fixed
  - Match Wordpress-Extra coding standards.
  - Fix text domain in i18n (make static instead of variable).
  - Fix domains not appearing properly if you have more than 20

## 1.0.2 - 2015-11-26
### Added
 - Developer hooks to filter api values

 ### Fixed
  - Update to WordPress coding standards

## 1.0.1 - 2015-11-23
### Added
 - Link to settings from the plugin page

### Fixed
 - Internationalization not enabled.

## 1.0.0 - 2015-11-09
- The initial release of this plugin


assets/js/admin.js
jQuery(document).ready(function($){

	// Initialize the bandwidth doughnut chart
	var bwChartID = $("#cmd-acf-bwchart").get(0).getContext("2d");
	var bwChart = new Chart(bwChartID).Doughnut(cmd_afc_bandwidth, {
		animation: false,
		tooltipTemplate : function(data) {
			return data.label + ': ' + get_bw_formatted_value(data);
		}
	} );

	// Initialize the ssl doughnut chart
	var sslChartID = $("#cmd-acf-sslchart").get(0).getContext("2d");
	var sslChart = new Chart(sslChartID).Doughnut(cmd_afc_ssl, {
		animation: false
	} );

	// Initialize the content type pie chart
	var ctChartID = $("#cmd-acf-ctchart").get(0).getContext("2d");
	var ctChart = new Chart(ctChartID).Pie(cmd_afc_content_types, {
		animation: false
	} );

	// Initialize the request by country pie chart
	var rcChartID = $("#cmd-acf-rcchart").get(0).getContext("2d");
	var rcChart = new Chart(rcChartID).Pie(cmd_afc_countries, {
		animation: false
	} );


	// Custom line chart definition
	Chart.types.Line.extend({
		name : "AltLine",

		initialize: function (data) {
			// If there are more than 7 datapoints then skip the label for every other one
			Chart.types.Line.prototype.initialize.apply(this, arguments);
			var xLabels = this.scale.xLabels;
			if ( xLabels.length > 7 ) {
				xLabels.forEach(function ( label, i ) {
					if ( 1 === i % 2 )
						xLabels[i] = '';
				});
			}
		}
	});

	// Initialize the main line chart
	var inChartID = $("#cmd-acf-linechart").get(0).getContext("2d");
	var inChart = new Chart(inChartID).AltLine(cmd_afc_interval, {
		animation: false,
		maintainAspectRatio: false,
		scaleBeginAtZero: true,
		bezierCurve : false,
		scaleLabel: function(data) {
				// Format the bandwidth label to MB
				if ( 'bandwidth' === cmd_afc_current_type ) {
					return ( ( data.value/1024 )/1024 ).toFixed(1) + 'MB';
				} else {
					return data.value;
				}
		},
		multiTooltipTemplate : function(data) {
				// Format the bandwidth tooltip to MB
				if ( 'bandwidth' === cmd_afc_current_type ) {
					return data.datasetLabel + ': ' + bytesToSize( data.value );
				} else {
					return data.datasetLabel + ': ' + data.value;
				}
		}
	});

	// Create the chart legend
	document.getElementById('cmd-acf-js-legend').innerHTML = inChart.generateLegend();

	// Grab the value of the display label to show
	function get_bw_formatted_value(data){
		var returnval = '';
		$.each(cmd_afc_bandwidth, function( i, item ) {
			if ( item.label == data.label ) {
				returnval = item.display;
			}
		});
	 	return returnval;
	}

	// Convert bytes to a more precice display
	function bytesToSize(bytes) {
		var sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
		if ( 0 === bytes ) {
			return '0 Byte';
		}
		var i = parseInt( Math.floor( Math.log(bytes) / Math.log(1024) ) );
		return ( bytes / Math.pow(1024, i) ).toFixed(2) + ' ' + sizes[i];
	}


});


analytics-for-cloudflare.php
<?php
/**
 * Plugin Name:       Analytics For Cloudflare
 * Plugin URI:        https://chuckmacdev.com
 * Description:       Access Cloudflare analytics information right from your WordPress dashboard.
 * Version:           1.1
 * Author:            ChuckMac Development
 * Author URI:        https://chuckmacdev.com
 * License:           GPL-2.0
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Text Domain:       cmd-analytics-for-cloudflare
 * Domain Path:       /languages
 */

// If this file is called directly, abort.
if ( ! defined( 'ABSPATH' ) ) {
	die;
}

/**
 * The core plugin class that is used to define internationalization,
 * admin-specific hooks, and public-facing site hooks.
 */
require plugin_dir_path( __FILE__ ) . 'includes/class-analytics-for-cloudflare.php';

/**
 * Begins execution of the plugin.
 *
 * Since everything within the plugin is registered via hooks,
 * then kicking off the plugin from this point in the file does
 * not affect the page life cycle.
 *
 * @since    1.0.0
 */
function run_cmd_analytics_for_cloudflare() {

	$plugin = new CMD_Analytics_For_Cloudflare();
	$plugin->set_base_file( __FILE__ );

}
run_cmd_analytics_for_cloudflare();

?>


includes/class-analytics-for-cloudflare.php
<?php
/**
 * The core plugin class.
 *
 * This is used to define internationalization, admin-specific hooks, and
 * public-facing site hooks.
 *
 * Also maintains the unique identifier of this plugin as well as the current
 * version of the plugin.
 *
 * @since      1.0.0
 * @package    CMD_Analytics_For_Cloudflare
 * @subpackage CMD_Analytics_For_Cloudflare/includes
 * @author     ChuckMac Development <chuck@chuckmacdev.com>
 */

// If this file is called directly, abort.
if ( ! defined( 'ABSPATH' ) ) {
	die;
}

class CMD_Analytics_For_Cloudflare {

	/** plugin version number */
	const VERSION = '1.1';

	/** plugin id */
	const PLUGIN_ID = 'cmd_analytics_for_cloudflare';

	/** plugin text domain */
	const TEXT_DOMAIN = 'cmd-analytics-for-cloudflare';

	/** plugin base file */
	public static $base_file;

	/**
	 * Define the core functionality of the plugin.
	 *
	 * Set the plugin name and the plugin version that can be used throughout the plugin.
	 * Load the dependencies, define the locale, and set the hooks for the admin area and
	 * the public-facing side of the site.
	 *
	 * @since    1.0.0
	 */
	public function __construct() {

		$this->set_locale();

		add_action( 'init', array( $this, 'init' ), 10 );

	}


	/**
	 * Initialize plugin - translations and classes
	 *
	 * @since 1.0.0
	 */
	public function init() {

		if ( is_admin() ) {
			// Admin Options Page
			require_once( 'class-analytics-for-cloudflare-admin-settings.php' );
			$admin = new CMD_Analytics_For_Cloudflare_Admin_Settings();

			// Admin Dashboard Widget
			require_once( 'class-analytics-for-cloudflare-admin-dashboard.php' );
			$admin = new CMD_Analytics_For_Cloudflare_Admin_Dashboard();
		}

	}

	/**
	 * Define the locale for this plugin for internationalization.
	 *
	 * Uses the Plugin_Name_i18n class in order to set the domain and to register the hook
	 * with WordPress.
	 *
	 * @since    1.0.0
	 * @access   private
	 */
	private function set_locale() {
		require_once( 'class-analytics-for-cloudflare-i18n.php' );
		$plugin_i18n = new CMD_Analytics_For_Cloudflare_i18n();
		$plugin_i18n->set_domain( 'cmd-analytics-for-cloudflare' );
		add_action( 'plugins_loaded', array( $plugin_i18n, 'load_plugin_textdomain' ) );
	}

	/**
	 * Render a template
	 *
	 * Allows parent/child themes to override the markup by placing the a file named basename( $default_template_path ) in their root folder,
	 * and also allows plugins or themes to override the markup by a filter. Themes might prefer that method if they place their templates
	 * in sub-directories to avoid cluttering the root folder. In both cases, the theme/plugin will have access to the variables so they can
	 * fully customize the output.
	 *
	 * @param  string $default_template_path The path to the template, relative to the plugin's `templates` folder
	 * @param  array  $variables             An array of variables to pass into the template's scope, indexed with the variable name so that it can be extract()-ed
	 * @param  string $require               'once' to use require_once() | 'always' to use require()
	 * @return string
	 */
	public static function render_template( $default_template_path, $variables = array(), $require = 'once' ) {
		$template_path = locate_template( basename( $default_template_path ) );
		if ( ! $template_path ) {
			$template_path = dirname( self::$base_file ) . '/templates/' . $default_template_path;
		}
		$template_path = apply_filters( CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_template_path', $template_path );

		if ( is_file( $template_path ) ) {
			extract( $variables );
			ob_start();

			if ( 'always' === $require ) {
				require( $template_path );
			} else {
				require_once( $template_path );
			}

			$template_content = apply_filters( CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_template_content', ob_get_clean(), $default_template_path, $template_path, $variables );
		} else {
			$template_content = false;
		}

		return $template_content;
	}



	/**
	 * Set the base filename of the plugin.
	 *
	 * @since    1.0.1
	 * @access   public
	 */
	public function set_base_file( $basefile ) {
		self::$base_file = $basefile;
	}
}


templates/admin/cmd-afc-dashboard-widget.php
<?php

/**
 * The display output of the dashboard widget
 *
 *
 * @since      1.0.0
 * @package    CMD_Analytics_For_Cloudflare_Admin_Dashboard
 * @subpackage CMD_Analytics_For_Cloudflare_Admin_Dashboard/includes
 * @author     ChuckMac Development <chuck@chuckmacdev.com>
 */

// If this file is called directly, abort.
if ( ! defined( 'ABSPATH' ) ) {
	die;
}

?>

<form id="cmd-analytics-for-cloudflare-dash" method="POST">
	<select id="<?php echo esc_attr( CMD_Analytics_For_Cloudflare::PLUGIN_ID );?>_dashboard_range" name="<?php echo esc_attr( CMD_Analytics_For_Cloudflare::PLUGIN_ID );?>_dashboard_range" onchange="this.form.submit()">
		<?php
		foreach ( $time_options as $key => $value ) {
			echo '<option value="' . esc_attr( $key ) . '"" ' . selected( $current_time, $key ) . '">' . esc_html( $value ) . '</option>';
		}
		?>
	</select>
	<select id="<?php echo esc_attr( CMD_Analytics_For_Cloudflare::PLUGIN_ID );?>_dashboard_type" name="<?php echo esc_attr( CMD_Analytics_For_Cloudflare::PLUGIN_ID );?>_dashboard_type" onchange="this.form.submit()">
		<?php
		foreach ( $display_options as $key => $value ) {
			echo '<option value="' . esc_attr( $key ) . '"" ' . selected( $current_type, $key ) . '">' . esc_html( $value ) . '</option>';
		}
		?>
	</select>
</form>

<div class="cmd-afc-wrapper">

	<h3 class="cmd-acf-heading"><?php echo esc_html( $display_options[ $current_type ] ); ?> / <?php echo esc_html( $time_options[ $current_time ] ); ?></h3>
	<canvas id="cmd-acf-linechart" class="line-chart" width="545" height="545"></canvas>
	<div id="cmd-acf-js-legend" class="chart-legend"></div>

	<div class="inside">
		<div class="small-box">
			<h3><?php esc_html_e( 'Total Requests', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->requests->all ) ? $analytics->totals->requests->all : '--' ) ); ?></p>
		</div>
		<div class="small-box">
			<h3><?php esc_html_e( 'Total Pageviews', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->pageviews->all ) ? $analytics->totals->pageviews->all : '--' ) ); ?></p>
		</div>
		<div class="small-box">
			<h3><?php esc_html_e( 'Total Unique Visitors', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->uniques->all ) ? $analytics->totals->uniques->all : '--' ) ); ?></p>
		</div>
		<div class="small-box">
			<h3><?php esc_html_e( 'Threats Detected', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->threats->all ) ? $analytics->totals->threats->all : '--' ) ); ?></p>
		</div>
		<div class="small-box">
			<h3><?php esc_html_e( 'Total Bandwidth', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->bandwidth->all ) ? CMD_Analytics_For_Cloudflare_Admin_Dashboard::format_bytes( $analytics->totals->bandwidth->all ) : '--' ) ); ?></p>
		</div>
		<div class="small-box">
			<h3><?php esc_html_e( 'Search Engine Crawls', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<p><?php echo esc_html( ( isset( $analytics->totals->pageviews->search_engine ) ? array_sum( (array) $analytics->totals->pageviews->search_engine ) : '--' ) ); ?></p>
		</div>
	</div>

	<div class="inside">
		<div class="donut-box">
			<h3><?php esc_html_e( 'Bandwidth', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<canvas id="cmd-acf-bwchart" class="donut-chart" width="400" height="400"></canvas>
		</div>
		<div class="donut-box">
			<h3><?php esc_html_e( 'SSL Traffic', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<canvas id="cmd-acf-sslchart" class="donut-chart" width="400" height="400"></canvas>
		</div>
		<div class="donut-box">
			<h3><?php esc_html_e( 'Content Types', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<canvas id="cmd-acf-ctchart" class="donut-chart" width="400" height="400"></canvas>
		</div>
		<div class="donut-box">
			<h3><?php esc_html_e( 'Requests by Country', 'cmd-analytics-for-cloudflare' ); ?></h3>
			<canvas id="cmd-acf-rcchart" class="donut-chart" width="400" height="400"></canvas>
		</div>
	</div>
</div>


includes/class-analytics-for-cloudflare-i18n.php
<?php

/**
 * Define the internationalization functionality.
 *
 * Loads and defines the internationalization files for this plugin
 * so that it is ready for translation.
 *
 * @since      1.0.0
 * @package    CMD_Analytics_For_Cloudflare_i18n
 * @subpackage CMD_Analytics_For_Cloudflare_i18n/includes
 * @author     ChuckMac Development <chuck@chuckmacdev.com>
 */

// If this file is called directly, abort.
if ( ! defined( 'ABSPATH' ) ) {
	die;
}

class CMD_Analytics_For_Cloudflare_i18n {

	/**
	 * The domain specified for this plugin.
	 *
	 * @since    1.0.0
	 * @access   private
	 * @var      string    $domain    The domain identifier for this plugin.
	 */
	private $domain;

	/**
	 * Load the plugin text domain for translation.
	 *
	 * @since    1.0.0
	 */
	public function load_plugin_textdomain() {

		load_plugin_textdomain(
			$this->domain,
			false,
			dirname( dirname( plugin_basename( __FILE__ ) ) ) . '/languages/'
		);

	}

	/**
	 * Set the domain equal to that of the specified domain.
	 *
	 * @since    1.0.0
	 * @param    string    $domain    The domain that represents the locale of this plugin.
	 */
	public function set_domain( $domain ) {
		$this->domain = $domain;
	}
}


includes/class-analytics-for-cloudflare-admin-dashboard.php
<?php

/**
 * Handle the creation of the admin dashboard widget.
 *
 *
 * @since      1.0.0
 * @package    CMD_Analytics_For_Cloudflare_Admin_Dashboard
 * @subpackage CMD_Analytics_For_Cloudflare_Admin_Dashboard/includes
 * @author     ChuckMac Development <chuck@chuckmacdev.com>
 */

// If this file is called directly, abort.
if ( ! defined( 'ABSPATH' ) ) {
	die;
}
class CMD_Analytics_For_Cloudflare_Admin_Dashboard {

	/** value of the plugin options field */
	private $plugin_options;

	/**
	 * Set the dynamic CloudFlare API information needed to build a reuquest.
	 *
	 * @since    1.0.0
	 */
	public function __construct() {

		add_action( 'wp_dashboard_setup', array( $this, 'register_dashboard_widget' ) );
		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_dashboard_scripts_styles' ) );

	}

	/**
	 * Register the CloudFlare for Analytics dashboard widget.
	 *
	 * @since    1.0.0
	 */
	public function register_dashboard_widget() {

		wp_add_dashboard_widget(
			CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_dashboard',
			__( 'Analytics For CloudFlare', 'cmd-analytics-for-cloudflare' ),
			array( $this, 'display_dashboard_widget' )
		);

	}

	/**
	 * Enqueue the necessary external javascript and css files needed.
	 *
	 * Only on the admin dashboard screen (index.php)
	 *
	 * @since    1.0.0
	 */
	public function enqueue_dashboard_scripts_styles( $hook ) {

		if ( 'index.php' === $hook ) {
			wp_enqueue_style( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-css', plugins_url( 'assets/css/admin.css' , dirname( __FILE__ ) ) );
			wp_enqueue_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-moment-js-lib', plugins_url( 'lib/Moment.js-2.10.6/moment.js' , dirname( __FILE__ ) ), array( 'jquery' ) );
			wp_enqueue_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-charts-js-lib', plugins_url( 'lib/Chart.js-1.0.2/Chart.min.js' , dirname( __FILE__ ) ), array( 'jquery', CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-moment-js-lib' ) );
		}

	}

	/**
	 * The main dashboard function.
	 *
	 * Query the CloudFlare API, output all the data to the dashboard widget.
	 *
	 * @since    1.0.0
	 */
	public function display_dashboard_widget() {

		$this->plugin_options = get_option( CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_settings' );

		// Available dropdown display items
		$time_options = array(
							#'-30'  =>  __('Last 30 minutes', 'cmd-analytics-for-cloudflare' ),  ## ENTERPRISE PLAN
							#'-360' => __('Last 6 hours', 'cmd-analytics-for-cloudflare' ),      ## PRO PLAN
							#'-720' => __('Last 12 hours', 'cmd-analytics-for-cloudflare' ),     ## PRO PLAN
							'-1440'  => __( 'Last 24 hours', 'cmd-analytics-for-cloudflare' ),
							'-10080' => __( 'Last week', 'cmd-analytics-for-cloudflare' ),
							'-43200' => __( 'Last month', 'cmd-analytics-for-cloudflare' ),
						);

		$display_options = array(
							'requests'  => __( 'Requests', 'cmd-analytics-for-cloudflare' ),
							'pageviews' => __( 'Pageviews', 'cmd-analytics-for-cloudflare' ),
							'uniques'   => __( 'Unique Visitors', 'cmd-analytics-for-cloudflare' ),
							'bandwidth' => __( 'Bandwidth', 'cmd-analytics-for-cloudflare' ),
						);

		//Default - last week / requests
		$current_time = '-10080';
		$current_type = 'pageviews';

		// Check to see if the user already has a view setup, if not use defaults - last week / requests
		$user_id = ( isset( $_REQUEST['cmd_analytics_for_cloudflare_current_user'] ) ) ? $_REQUEST['cmd_analytics_for_cloudflare_current_user'] : get_current_user_id();
		$user_view = get_user_meta( $user_id, '_cmd_analytics_for_cloudflare_view', true );

		$current_time = ( isset( $user_view['current_time'] ) ) ? $user_view['current_time'] : '-10080';
		$current_type = ( isset( $user_view['current_type'] ) ) ? $user_view['current_type'] : 'pageviews';

		$this->plugin_options = get_option( CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_settings' );

		//Check if the form was submitted to change the view
		if ( ( isset( $_REQUEST['cmd_analytics_for_cloudflare_dashboard_range'] ) ) && ( array_key_exists( $_REQUEST['cmd_analytics_for_cloudflare_dashboard_range'], $time_options ) ) ) {
			$current_time = $_REQUEST['cmd_analytics_for_cloudflare_dashboard_range'];
		}
		if ( ( isset( $_REQUEST['cmd_analytics_for_cloudflare_dashboard_type'] ) ) && ( array_key_exists( $_REQUEST['cmd_analytics_for_cloudflare_dashboard_type'], $display_options ) ) ) {
			$current_type = $_REQUEST['cmd_analytics_for_cloudflare_dashboard_type'];
		}

		// Save current user view
		update_user_meta( $user_id, '_cmd_analytics_for_cloudflare_view', array( 'current_time' => $current_time, 'current_type' => $current_type ) );

		//Set our caching options
		$options = get_option( CMD_Analytics_For_Cloudflare::PLUGIN_ID . '_settings' );
		$cache_time = ( isset( $this->plugin_options['cache_time'] ) ? $this->plugin_options['cache_time'] : '900' );

		//Check our transiant for the analytics object for our current view, if not found then pull the data
		if ( '0' === $cache_time || false === ( $analytics = get_transient( 'cmd_afc_results_' . $current_time . '_' . $current_type ) ) ) {

			require_once( 'class-analytics-for-cloudflare-api.php' );
			$cloudflare = new CMD_Analytics_For_Cloudflare_Api();
			$analytics = $cloudflare->get_analytics( array( 'since' => $current_time ) );

			//If we encounter an error, show it and don't cache
			if ( is_wp_error( $analytics ) ) {
				echo '<h3 class=>' . esc_html__( 'Unable to connect to CloudFlare', 'cmd-analytics-for-cloudflare' ) . '</h3>';
				echo '<p>' . esc_html( $analytics->get_error_message() ) . '</p>';
				echo '<p>' . esc_html__( sprintf( 'View the %sCloudFlare For Analytics settings%s', '<a href="' . admin_url( 'options-general.php?page=cmd_analytics_for_cloudflare' ) . '">', '</a>' ), 'cmd-analytics-for-cloudflare' ) . '</p>';
				return;
			}

			//Set the transient cache for the current view
			set_transient( 'cmd_afc_results_' . $current_time . '_' . $current_type, $analytics, $cache_time );

		}

		// Initialize all the javascript data
		$this->parse_analytics_to_js( $analytics, $current_type, $current_time, $display_options );

		do_action( 'cmd_analytics_for_cloudflare_before_dashboard' );

		// Render the display from the dashboard template file
		echo CMD_Analytics_For_Cloudflare::render_template( 'admin/cmd-afc-dashboard-widget.php',
			array(
				'time_options'    => $time_options,
				'current_time'    => $current_time,
				'display_options' => $display_options,
				'current_type'    => $current_type,
				'analytics'       => $analytics,
			)
		);

		do_action( 'cmd_analytics_for_cloudflare_after_dashboard' );
	}


	/**
	 * Convert the analytics return data to javascript objects.
	 *
	 * @since    1.0.0
	 */
	public function parse_analytics_to_js( $analytics, $current_type, $current_time, $display_options ) {

		$bandwidth = array(
			'0' => array(
				'value'     => $analytics->totals->bandwidth->cached,
				'color'     => '#F68B1F',
				'highlight' => '#F4690C',
				'label'     => __( 'Cached', 'cmd-analytics-for-cloudflare' ),
				'display'   => $this->format_bytes( $analytics->totals->bandwidth->cached ),
			),
			'1' => array(
				'value'     => $analytics->totals->bandwidth->uncached,
				'color'     => '#A9A9A9',
				'highlight' => '#8F9CA8',
				'label'     => __( 'Uncached', 'cmd-analytics-for-cloudflare' ),
				'display'   => $this->format_bytes( $analytics->totals->bandwidth->uncached ),
			),
		);

		$ssl = array(
			'0' => array(
				'value'     => $analytics->totals->requests->ssl->encrypted,
				'color'     => '#F68B1F',
				'highlight' => '#F4690C',
				'label'     => __( 'Encypted', 'cmd-analytics-for-cloudflare' ),
			),
			'1' => array(
				'value'     => $analytics->totals->requests->ssl->unencrypted,
				'color'     => '#A9A9A9',
				'highlight' => '#8F9CA8',
				'label'     => __( 'Unencrypted', 'cmd-analytics-for-cloudflare' ),
			),
		);

		$colors = apply_filters( 'cmd_analytics_for_cloudflare_chart_colors', array( '#F68B1F', '#4D4D4D', '#5DA5DA', '#60BD68', '#F17CB0', '#B2912F', '#B276B2', '#9BFFE4', '#DECF3F', '#F15854' ) );

		// Create the content type totals, sort by number of requests
		$totals = (array) $analytics->totals->requests->content_type;
		arsort( $totals );
		$i = 0;
		$content_types = array();
		foreach ( $totals as $type => $value ) {
			$content_types[] = array(
				'value' => $value,
				'label' => $type,
				'color' => $colors[ $i ],
			);
			$i++;
			if ( $i > 9 ) {
				$i = 0;
			}
		}

		// Create the country totals, sort by number of requests
		$totals = (array) $analytics->totals->requests->country;
		arsort( $totals );
		$i = 0;
		$countries = array();
		foreach ( $totals as $type => $value ) {
			$countries[] = array(
				'value' => $value,
				'label' => $type,
				'color' => $colors[ $i ],
			);
			$i++;
			if ( $i > 9 ) {
				$i = 0;
			}
		}

		// Configuration for the main line chart
		$interval_chart = array();
		if ( ( 'requests' === $current_type ) || ( 'bandwidth' === $current_type )  ) {
			$interval_chart['datasets'][0] = array(
				'label' => $display_options[ $current_type ],
				'color' => 'rgba(76,255,0,0.2)',
				'fillColor' => 'rgba(76,255,0,0.2)',
				'strokeColor' => 'rgba(63,211,0,1)',
				'scaleFontSize' => '0',
				'pointColor' => 'rgba(50,168,0,1)',
				'pointStrokeColor' => '#fff',
				'pointHighlightFill' => '#fff',
				'pointHighlightStroke' => 'rgba(220,220,220,1)',
			);
			$interval_chart['datasets'][1] = array(
				'label' => __( 'Cached', 'cmd-analytics-for-cloudflare' ),
				'fillColor' => 'rgba(246,139,31,0.2)',
				'strokeColor' => 'rgba(234,101,0,1)',
				'scaleFontSize' => '0',
				'pointColor' => 'rgba(232,171,127,1)',
				'pointStrokeColor' => '#fff',
				'pointHighlightFill' => '#fff',
				'pointHighlightStroke' => 'rgba(220,220,220,1)',
			);
			$interval_chart['datasets'][2] = array(
				'label' => __( 'Uncached', 'cmd-analytics-for-cloudflare' ),
				'fillColor' => 'rgba(129,129,129,0.2)',
				'strokeColor' => 'rgba(143,156,168,1)',
				'scaleFontSize' => '0',
				'pointColor' => 'rgba(118,128,137,1)',
				'pointStrokeColor' => '#fff',
				'pointHighlightFill' => '#fff',
				'pointHighlightStroke' => 'rgba(220,220,220,1)',
			);
		} else {
			$interval_chart['datasets'][0] = array(
				'label' => __( 'Pageviews', 'cmd-analytics-for-cloudflare' ),
				'fillColor' => 'rgba(246,139,31,0.2)',
				'strokeColor' => 'rgba(234,101,0,1)',
				'scaleFontSize' => '0',
				'pointColor' => 'rgba(232,171,127,1)',
				'pointStrokeColor' => 'fff',
				'pointHighlightFill' => '#fff',
				'pointHighlightStroke' => 'rgba(220,220,220,1)',
			);
		}

		foreach ( $analytics->timeseries as $interval ) {

			// Set the date format for the chart (hours or dates)
			if ( '-1440' === $current_time ) {
				$interval_chart['labels'][] = apply_filters( 'cmd_analytics_for_cloudflare_interval_dateformat', get_date_from_gmt( date( 'Y-m-d H:i:s', strtotime( $interval->since ) ), 'ga' ), $interval->since, $current_time );
			} else {
				$interval_chart['labels'][] = apply_filters( 'cmd_analytics_for_cloudflare_interval_dateformat', get_date_from_gmt( date( 'Y-m-d H:i:s', strtotime( $interval->since ) ), 'm/d' ), $interval->since, $current_time );
			}

			if ( ( 'requests' === $current_type ) || ( 'bandwidth' === $current_type ) ) {
				$interval_chart['datasets'][0]['data'][] = $interval->$current_type->all;
				$interval_chart['datasets'][1]['data'][] = $interval->$current_type->cached;
				$interval_chart['datasets'][2]['data'][] = $interval->$current_type->uncached;
			} else {
				$interval_chart['datasets'][0]['data'][] = $interval->$current_type->all;
			}
		}

		$ajax = array(
			'ajax_url'      => admin_url( 'admin-ajax.php' ),
			'ajax_nonce'    => wp_create_nonce( 'cmd_afc_nonce' ),
			'cmd_plugin_id' => CMD_Analytics_For_Cloudflare::PLUGIN_ID,
			'current_user'  => get_current_user_id(),
		);

		// Register and localize all the script data
		wp_register_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', plugins_url( 'assets/js/admin.js' , dirname( __FILE__ ) ), array( 'jquery' ), CMD_Analytics_For_Cloudflare::VERSION, true );

		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_interval', $interval_chart );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_bandwidth', $bandwidth );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_ssl', $ssl );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_content_types', $content_types );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_countries', $countries );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_current_type', $current_type );
		wp_localize_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js', 'cmd_afc_ajax', $ajax );

		wp_enqueue_script( CMD_Analytics_For_Cloudflare::TEXT_DOMAIN . '-js' );

	}


	/**
	 * Convert bytes to more consise values depending on the size.
	 *
	 * Add a suffix to the end to denote the converted value
	 *   B  = bytes
	 *   KB = kilobytes
	 *   MB = megabytes
	 *   GB = gigabytes
	 *   TB = terrabytes
	 *
	 * @since     1.0.0
	 * @param     string   $bytes       A numeric value representing a byte count
	 * @param     string   $precision   Number of decimal places to return
	 * @return    string   $return      The formatted value of bytes
	 */
	public static function format_bytes( $bytes, $precision = 2 ) {
		$units = array( 'B', 'KB', 'MB', 'GB', 'TB' );

		$bytes = max( $bytes, 0 );
		$pow = floor( ( $bytes ? log( $bytes ) : 0 ) / log( 1024 ) );
		$pow = min( $pow, count( $units ) - 1 );

		$bytes /= pow( 1024, $pow );

		return round( $bytes, $precision ) . ' ' . $units[ $pow ];
	}
}
